[Week 7] 인가(Authorization) 구현하기

yhyojoo opened the pull request:
# 과제 목표
지금은 로그인을 했다고 하더라도, 다른 사람의 정보를 내 마음대로 수정할 수 있습니다. 내 정보는 오직 나만 수정할 수 있어야 합니다. 오직 나만이 내 정보를 수정할 수 있도록 Spring Security를 이용해서 구현해 주세요.

현재 사용자의 비밀번호를 평문으로 저장하고 있습니다. 누군가 외부에서 침투를 하거나 혹은 내부에서 비밀번호를 탈취를 하더라도 원래 비밀번호를 알아낼 수 없도록 비밀번호를 암호화하여 저장해 주세요.

## 로그인이 필요없는 API
>로그인 - `POST /session`
>회원 생성하기 - `POST /users`
>고양이 장난감 목록 얻기 - `GET /products`
>고양이 장난감 상세 조회하기 - `GET /products/{id}`
## 로그인이 필요한 API
>고양이 장난감 등록하기 - `POST /products`
>고양이 장난감 수정하기 - `PATCH /products/{id}`
>고양이 장난감 삭제하기 - `DELETE /products/{id}`
>회원 수정하기 - `POST /users/{id}`
>회원 삭제하기 - `DELETE /users/{id}`

yhyojoo committed:
시작!

yhyojoo commented:
안녕하세요! 한 주간 잘 부탁드립니다 😊

yhyojoo committed:
SecurityJavaConfig 생성

yhyojoo committed:
SecurityJavaConfig에 AuthenticationFilter 추가

사용자 인증 처리를 위해 authenticationManager를 파라미터로 받는 AuthenticationFilter를 생성하고 SecurityJavaConfig에 추가

yhyojoo committed:
인터셉트의 인증 구현 부분을 JwtAuthenticationFilter로 이동

인터셉트가 하던 역할을 SpringSecurity에서 처리할 수 있도록 인터셉트의
인증 구현 부분을 JwtAuthenticationFilter로 이동

yhyojoo committed:
JwtAuthenticationFilter에서 빈 토큰일 경우 예외를 던질 수 있도록 코드 수정

yhyojoo committed:
JwtAuthenticationFilter에서 예외를 그대로 처리하도록 코드 수정

yhyojoo committed:
SecurityJavaConfig에 우선 실행될 AuthenticationErrorFilter 추가

yhyojoo committed:
AuthenticationErrorFilter에 예외 처리 코드 작성

yhyojoo committed:
WebJavaConfig, AuthenticationInterceptor 삭제

yhyojoo committed:
ControllerErrorAdvice에서 handleInvalidAccessToken() 제거

yhyojoo committed:
사용자 인증 정보 관리 기능 추가

yhyojoo committed:
Authorization이 null이 아닌 경우를 처리하는 코드로 수정

JwtAuthenticationFilter의 doFilterInternal 메소드 내 Authorization이
null인 경우를 처리하는 부분을 null이 아닌 경우를 처리하는 코드로 수정

yhyojoo committed:
어노테이션 기반 보안을 적용하기 위해 @EnableGlobalMethodSecurity 추가

prePostEnabled 속성을 true로 설정해 @PreAuthorize 어노테이션을 사용하여
인가 처리를 할 수 있도록 함

yhyojoo committed:
사용자 인증이 필요한 부분에 @PreAuthorize 추가

yhyojoo committed:
상품 삭제 시 유효하지 않은 토큰에 대한 테스트 코드 작성

yhyojoo committed:
사용자 권한 확인 코드 추가

yhyojoo committed:
세션을 사용하지 않기 위해 Config에 SessionManagement() 추가

yhyojoo committed:
사용되지 않는 코드 삭제

johngrib reviewed (commented):
현재 컴파일이 되지 않습니다. 컴파일이 되는 상태로 작업해 올려주시길 바랍니다.

이번주 부터는 커밋 메시지도 리뷰를 합니다. 커밋할 때 커밋 메시지 타이틀과 본문을 어떻게 작성할지 고민해 보시고 열심히 작성해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,14 @@
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+         http
+                 .csrf().disable();
```
인덴트가 너무 깊습니다.

johngrib reviewed (commented):
d95c044816b91cc4be67f0359a39b9f6d5282a42 커밋 메시지 `SecurityJavaConfig에 authenticationFilter 추가`의 의미를 풀어서 설명할 수 있나요?

johngrib reviewed (commented):
c91ba1193da641a4f74371746b889d732e177fbf `인터셉트의 인증 구현 부분을 JwtAuthenticationFilter로 이동`이라고 커밋 메시지 타이틀을 작성하셨는데요, 그 이유를 커밋 메시지 본문에 적어주시면 더 좋을 것입니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -33,14 +33,13 @@ protected void doFilterInternal(HttpServletRequest request,
+            }
 
-        String accessToken = authorization.substring("Bearer ".length());
+            String accessToken = authorization.substring("Bearer ".length());
```
4bbfa55b1636725bf845a20eec82a025ba683a18 커밋 메시지는 `JwtAuthenticationFilter에서 빈 토큰일 경우 예외를 던질 수 있도록 코드 수정` 인데요, 커밋 메시지와 관계 없는 작업이 포함되어 있습니다.

yhyojoo replied:
앞으로 주의하도록 하겠습니다!

johngrib reviewed (commented):
a3bc7049c352d921f9e7ec6c535fce7363076e5b `JwtAuthenticationFilter에서 예외를 그대로 처리하도록 코드 수정` 왜 이런 결정을 했는지에 대해서도 커밋 메시지 본문에 작성해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/AuthenticationErrorFilter.java:
```diff
@@ -0,0 +1,15 @@
+
+public class AuthenticationErrorFilter extends HttpFilter {
+    @Override
+    protected void doFilter(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
```
한 줄이 너무 깁니다. 적절히 개행해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/AuthenticationErrorFilter.java:
```diff
@@ -0,0 +1,15 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class AuthenticationErrorFilter extends HttpFilter {
```
JavaDoc 주석을 작성해 주시면 좋겠습니다.

johngrib reviewed (commented):
ae495c27bf4854a9c332803a772d51a9556d1669  `null이 아닌 경우를 처리하는 코드로 수정` 커밋 메시지가 애매합니다. 어느 부분을 어떻게 고쳤다는 의미일까요?

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -5,17 +5,18 @@
 import javax.servlet.Filter;
 
 @Configuration
+@EnableGlobalMethodSecurity(prePostEnabled = true)
```
이 애노테이션이 어떤 의미를 갖는지 커밋 메시지에서 설명하고 있었으면 좋았을 거에요.

yhyojoo committed:
비밀번호 암호화 기능 추가

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -303,4 +303,25 @@ void deleteWithNotExistedProduct() throws Exception {
+    void deleteWithInvalidAccessToken() throws Exception {
+        mockMvc.perform(
+                delete("/products/1")
+                        .accept(MediaType.APPLICATION_JSON_UTF8)
```
Deprecated 된 값일 거에요.

yhyojoo committed:
PasswordEncoder Bean 추가

yhyojoo committed:
서비스에 PasswordEncoder 의존성 주입

yhyojoo committed:
테스트에 비밀번호 암호화 코드 추가

yhyojoo committed:
순환 참조 문제를 해결하기 위해 PasswordEncoder Bean 이동

SecurityJavaConfig에서 App으로 Bean 이동

yhyojoo committed:
사용되지 않는 코드 제거

yhyojoo committed:
AuthenticationErrorFilter에 JavaDoc 주석 작성

johngrib reviewed (commented):
아직 몇몇 클래스 파일이 제외되어 있어 컴파일이 안 되는 상태입니다.

johngrib reviewed (commented):
858df3586268aa0d0edd66819a10d4e5ce4dc186 `순환 참조 문제를 해결하기 위해 PasswordEncoder Bean 이동` 어디로 이동했는지에 대해서도 커밋 메시지 본문에 작성해주셨으면 더 좋았을 거에요.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -169,7 +169,6 @@ void createWithValidAttributes() throws Exception {
     void createWithInvalidAttributes() throws Exception {
         mockMvc.perform(
                 post("/products")
-                        .accept(MediaType.APPLICATION_JSON_UTF8)
```
이걸 삭제한 대신 `MediaType.APPLICATION_JSON`를 추가해주셨다면 좋았을 것입니다.

![image](https://user-images.githubusercontent.com/1855714/110939222-6e9c8e00-8378-11eb-922a-7b381cfe7dc1.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/AuthenticationErrorFilter.java:
```diff
@@ -10,6 +10,9 @@
 import java.io.IOException;
 
+/**
+ * 사용자 인증 에러 처리 필터
```
JavaDoc description 마지막에는 마침표를 찍어 주셔야 해요. `사용자 인증 에러 처리 필터` -> `사용자 인증 에러 처리 필터.`

yhyojoo committed:
컴파일 에러 해결

yhyojoo committed:
사용되지 않는 코드 제거

yhyojoo committed:
ProductTest 파일 변경

yhyojoo committed:
JavaDoc에 마침표 추가

johngrib reviewed (commented):
ecda7601ca8750c923ba08135095234642982efd 커밋 메시지 타이틀을 `컴파일 에러를 해결`이라고 했다면 더 좋았을 거에요. 그냥 `파일 추가`는 의도가 모호하죠.

![image](https://user-images.githubusercontent.com/1855714/111029275-45930080-843f-11eb-8102-082287f03d53.png)

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/SessionControllerTest.java:
```diff
@@ -21,52 +26,77 @@
+    private SessionRequestData validLoginData;
+    private SessionRequestData invalidLoginData;
+
+    private String VALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOjF9.ZZ3CUl0jxeLGvQ1Js5nG2Ty5qGTlqai5ubDMXZOdaDk";;
```
세미콜론이 두 개 있네요?
```suggestion
    private String VALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VySWQiOjF9.ZZ3CUl0jxeLGvQ1Js5nG2Ty5qGTlqai5ubDMXZOdaDk";
```

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,63 @@
+        }
+
+        String method = request.getMethod();
+        if (method.equals("GET")) {
+            return true;
+        }
+        return false;
```
이렇게 하면 어떨까요?
```suggestion
        return "GET".equals(request.getMethod());
```

johngrib reviewed (commented):
벌써 일요일이 되었네요. 회고를 작성해 공유해 주세요.

yhyojoo committed:
오타 수정

yhyojoo committed:
테스트 통과를 위해 filterWithPathAndMethod 메소드 수정

JwtAuthenticationFilter의 filterWithPathAndMethod 메소드를 두 개의 메소드 filterWithPath/filterWithMethod로 분리해 요청에 대한 필터 작업을 수행하도록 함

yhyojoo committed:
불필요한 줄 제거

yhyojoo committed:
불필요한 코드 제거

johngrib reviewed (commented):
b346a9c0ec4bd28dddd2492b455e648a54dd7629 이 커밋 메시지는 설명이 더 필요해 보입니다.
![image](https://user-images.githubusercontent.com/1855714/111149675-82dac800-85d0-11eb-8fba-700547856d44.png)

johngrib reviewed (commented):
774a711bc7b24449a95828254317de854b0d970f 이렇게 `@`를 쓰게 되면 이 아이디를 가진 다른 github 사용자가 링크가 되고, 그 사람에게 알림이 가게 돼요. 그러면 그 사람이 여기로 와서 효주님의 PR을 구경하고 갈 수도 있겠죠?

![image](https://user-images.githubusercontent.com/1855714/111150094-0694b480-85d1-11eb-8eb9-8272331995a3.png)

이런 일을 방지하기 위해 `@`을 쓰게 되면 양쪽에 백틱 ``` ` ``` 을 추가해 주도록 합니다.
다음과 같이 작성하면 됩니다.
```
어노테이션 기반 보안을 적용하기 위해 `@EnableGlobalMethodSecurity` 추가

prePostEnabled 속성을 true로 설정해 `@PreAuthorize` 어노테이션을 사용하여
인가 처리를 할 수 있도록 함
```

johngrib reviewed (commented):
c64d14c11dc01197d4cad14d286496f72f88c6bc 커밋 메시지와 작업 내용이 많이 다릅니다.

![image](https://user-images.githubusercontent.com/1855714/111150472-799e2b00-85d1-11eb-9ac3-e8a3cade8b38.png)

johngrib reviewed (commented):
한 주간 고생 많으셨습니다.
